Fix InvalidArgumentError error when mixed precision policy is used in Attention/AdditiveAttention layer

yongtang opened the pull request:
This PR tries to address the issue raised in #46064 where
InvalidArgumentError error is thrown when mixed precision policy is used
in keras Attention/AdditiveAttention layer.

This PR fixes #46064.

This PR also fixes  #43261.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Fix InvalidArgumentError error when mixed precision plicy is used in Attention/AdditiveAttention layer

This PR tries to address the issue raised in 46064 where
InvalidArgumentError error is thrown when mixed precision plicy is used
in keras Attention/AdditiveAttention layer.

This PR fixes 46064.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@gbaned Any update on the review?

fchollet commented on tensorflow/python/keras/layers/dense_attention_test.py:
```diff
@@ -757,6 +760,19 @@ def test_serialization(self):
     new_layer = dense_attention.AdditiveAttention.from_config(config)
     self.assertEqual(new_layer.use_scale, True)
 
+  def test_mixed_float16_policy(self):
```
Need input from @reedwm to see if this is an appropriate way to test mixed precision behavior.

LGTM otherwise!

reedwm replied:
Can you also add a test to [layer_correctness_test.py](https://github.com/tensorflow/tensorflow/blob/4133fef6917fbf973229a8a8047ba2056cd8b8ee/tensorflow/python/keras/mixed_precision/layer_correctness_test.py#L75) both for Attention and AdditiveAttention? Just need to add a parameter for both classes to the parameterized test. This automatically tests the outputs and gradients are the same between mixed precision and float32.

Also better to use the internal-only [policy_scope](https://github.com/tensorflow/tensorflow/blob/7d43911faa7bae5c56e3ffe4bd3c48e75ba37e5c/tensorflow/python/keras/mixed_precision/policy.py#L530) function to set the policy, which is equivalent to the try-finally statement used here

yongtang replied:
Thanks @fchollet, the PR has been updated with comments addressed.

yongtang committed:
Update to use policy.policy_scope to address review comment.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @reedwm. The PR has been updated to replace with `policy.policy_scope`.

For adding tests to `layer_correctness_test.py`, when I add the following:
```
      ('AdditiveAttentionLayerCausal',
       lambda: dense_attention.AdditiveAttention(causal=True),
       [(2, 3, 4), (2, 3, 4), (2, 3, 4)]),
```

The model's output are all `nan`s. Not sure if this is a known issue?

reedwm commented:
> The model's output are all `nan`s. Not sure if this is a known issue?

The issue is this line overflows:

https://github.com/tensorflow/tensorflow/blob/102e1f98552c4fe32a9abbb82ed22275f1be161b/tensorflow/python/keras/layers/dense_attention.py#L129

In float16, the `1.e9` overflows a float16 value, than when multiplied by zero, becomes NaN (Inf multiplied by 0 is NaN). To fix, if `padding_mask` is float16, use 1.e4 instead, with a comment stating it's to avoid overflow, then add the `layer_correctness_test.py` test.

Note `scores` may underflow but this is OK: softmax treats -Inf values as effectively masked out, which is exactly what the code is trying to do.

reedwm commented:
Actually, instead of replacing 1.e9 with 1.e4, replace 1.e9 with 65504, which is the maximum float16 value.

yongtang committed:
Add test in layer_correctness_test.py, and use 65504 for float16 padding_mask in Attention.

Update: additionally passing causal=True to Attention as well (from comment feedback)

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @reedwm @fchollet for the help! The PR has been updated. Please take a look.

reedwm commented on tensorflow/python/keras/mixed_precision/layer_correctness_test.py:
```diff
@@ -139,6 +139,11 @@ def _create_model_from_layer(self, layer, input_shapes):
        (2, 2, 2)),
       ('Bidirectional',
        lambda: wrappers.Bidirectional(recurrent.SimpleRNN(units=4)), (2, 2, 2)),
+      ('AttentionLayer', dense_attention.Attention,
```
Pass `causal=True` here, as in the next test. Since there already is an `Attention` test above that runs without causal.

yongtang replied:
Thanks @reedwm, updated.

gbaned commented:
@yongtang  Can you please check @reedwm's comments and keep us posted ? Thanks!

yongtang commented:
Thanks @reedwm for the help. The PR has been updated with all comments addressed. Please take a look.
